### PR TITLE
Revamped code for hamiltonian decomposition

### PR DESCRIPTION
**Pull request by nvitucci**
**Context:** Improve and finalize the code for decomposing hamiltonians.

**Description of the Change:** Updated the code and added more tests to existing branch.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -65,31 +68,77 @@
...
+                    assert all(np.allclose(o[0].matrix, o[1].matrix) for o in obs)
+
+                    linear_comb += decomposed_coeff[term] * functools.reduce(
+                        np.kron, [o.matrix for o in decomposed_obs[term].obs]
+                    )
+                else:
+                    assert np.allclose(decomposed_obs[term].matrix, expected_obs[term].matrix)
+                    linear_comb += decomposed_coeff[term] * decomposed_obs[term].matrix
+
+            assert np.allclose(test, linear_comb)
```
Comment (trbromley) [inline tests/test_utils.py:141]: My suggestion for the tests:

- We can keep things simple and have just a list of input test Hamiltonians, i.e., without the expected coefficients and observables. A permutation of coeffs and observables would still be a valid decomposition so it makes me think it's best not to define the correct/expected result.
- We have
```python
H_decomposed = sum([decomposed_coeff[i] * o.matrix for i, o in enumerate(decomposed_obs)])
assert np.allclose(test, H_decomposed)
```
- Additional test that the length of the coefficients and observables is as expected (i.e. log2 the length of the test H)
- Another test of the type of the observables:
```python
allowed_obs = (Tensor, Identity, PauliX, PauliY, PauliZ)
assert all([isinstance(o, allowed_obs) for o in decomposed_obs])
```
- A test that checks when `hide_identity=True`
- Two trivial-sounding tests for the sake of ensuring good coverage: tests designed to raise the two `ValueError` checks using
```python
with pytest.raises(TypeError, match="error message"):
```

I recommend having all these tests as separate methods within `TestDecomposition`.

File: tests/test_utils.py [hunk 2]
```diff
@@ -65,31 +68,77 @@
...
+        """Tests that decompose_hamiltonian successfully decomposes Hamiltonians into a
+        linear combination of Pauli matrices"""
+        for test, expected in test_hamiltonians:
+            decomposed = pu.decompose_hamiltonian(test)
+
+            decomposed_coeff, decomposed_obs = decomposed
+            expected_coeff, expected_obs = expected
+
+            assert len(decomposed_obs) == len(decomposed_coeff)
+            assert len(expected_obs) == len(expected_coeff)
```
[the remark this answers is not in the document]
Comment (nvitucci) [reply tests/test_utils.py:120]: True, I wanted to make explicit the fact that there have to be as many coefficients as there are terms, but your suggestion below seems more clear.

File: tests/test_utils.py [hunk 3]
```diff
@@ -65,31 +68,77 @@
...
-                    functools.reduce(operator.matmul, [t(i) for i, t in enumerate(base)]).matrix
-                )
-        for term in zip(coeff, matrices):
-            result = result + term[0] * term[1]
-        assert (H == result).all
+    @pytest.mark.parametrize("H", test_hamiltonians)
+    def test_decomposition(self, H):
+        """Tests that decompose_hamiltonian successfully decomposes Hamiltonians into a
+        linear combination of Pauli matrices"""
+        for test, expected in test_hamiltonians:
```
[the remark this answers is not in the document]
Comment (nvitucci) [reply tests/test_utils.py:113]: How could I not see this? Good catch :)

File: pennylane/utils.py [hunk 4]
```diff
@@ -31,11 +31,14 @@
...
-def decompose_hamiltonian(H):
+def decompose_hamiltonian(H, hide_identity=False):
     """Decomposes a Hermitian matrix into a linear combination of Pauli operators.
 
     Args:
-        H (array[complex]): an Hermitian matrix of dimension :math:`2^N\times 2^N`
+        H (array[complex]): an Hermitian matrix of dimension :math:`2^n\times 2^n`
+
+    Keyword Args:
+        hide_identity (bool): always show ~.Identity observables in the results
```
[the remark this answers is not in the document]
Comment (nvitucci) [reply pennylane/utils.py]: Makes sense.

File: pennylane/utils.py [hunk 5]
```diff
@@ -87,7 +94,12 @@ def decompose_hamiltonian(H):
             if not all(t is qml.Identity for t in term):
                 obs.append(
                     functools.reduce(
-                        matmul, [t(i) for i, t in enumerate(term) if t is not qml.Identity]
+                        matmul,
+                        [
+                            t(i)
+                            for i, t in enumerate(term)
+                            if t is not qml.Identity or not hide_identity
+                        ],
```
[the remark this answers is not in the document]
Comment (nvitucci) [reply pennylane/utils.py:103]: Yes, I refactored it a little, so maybe now it is even more clear.

File: tests/test_utils.py [hunk 6]
```diff
@@ -65,31 +69,88 @@
...
+                    np.kron, [o.matrix for o in decomposed_obs[term].obs]
                 )
-        for term in zip(coeff, matrices):
-            result = result + term[0] * term[1]
-        assert (H == result).all
+            else:
+                assert np.allclose(decomposed_obs[term].matrix, expected_obs[term].matrix)
+                linear_comb += decomposed_coeff[term] * decomposed_obs[term].matrix
+
+        assert np.allclose(test, linear_comb)
```
Comment (josh146) [inline tests/test_utils.py:153]: Very thorough testing :) I agree with @trbromley's comment about splitting up this particular tests into separate tests for readability and maintainability

File: pennylane/utils.py [hunk 7]
```diff
@@ -65,29 +70,33 @@ def decompose_hamiltonian(H):
...
 
-    This Hamiltonian can then be used in defining VQE problems using :class:`~VQECost`.
+    This Hamiltonian can then be used in defining VQE problems using :class:`~.VQECost`.
     """
-    N = int(np.log2(len(H)))
+    n = int(np.log2(len(H)))
+    N = 2 ** n
+
+    if len(H) - N != 0:
+        raise ValueError("Hamiltonian should have shape (2^n, 2^n), for any qubit number n>=1")
```
[the remark this answers is not in the document]
Comment (nvitucci) [reply pennylane/utils.py:79]: Good idea. This will also avoid parsing problems with `pytest.raises`.

[the remark this answers is not in the document]
Comment (nvitucci) [reply pennylane/utils.py]: Actually escaping is still necessary, but this anyway is more readable.

File: tests/test_utils.py (hunk 6 above)
Comment (nvitucci) [reply tests/test_utils.py:153]: Sounds good, I'll refactor as you both suggested.

File: tests/test_utils.py (hunk 1 above)
Comment (nvitucci) [reply tests/test_utils.py]: It should be better now. I only had one doubt:

> Additional test that the length of the coefficients and observables is as expected (i.e. log2 the length of the test H)

I assumed it is on the length of each observable when it's a tensor, as we don't return all the terms (e.g. when the coefficient is 0); if we did, the length would be 4**n anyway. Did I understand correctly?

**Comment by nvitucci** [comment]
> Don't forget to also add your name (and an entry regarding this PR) to the changelog @nvitucci!

I think that, before I can do this, the master branch has to be merged?

File: tests/test_utils.py (hunk 1 above)
Comment (trbromley) [reply tests/test_utils.py]: Nice! Actually I was thinking the other way (i.e., 4**n) but I think what you did is better (and makes sense)!

**Comment by trbromley** [comment]
Thanks @nvitucci, I just left a few comments. Looking good!

Think we're nearly there now, feel free to update the changelog (I merged `master` into `decompose-hamiltonian`).

File: tests/test_utils.py [hunk 8]
```diff
@@ -65,31 +69,64 @@
...
+            all_identities = all(isinstance(o, Identity) for o in tensor.obs)
+            no_identities = not any(isinstance(o, Identity) for o in tensor.obs)
+            assert all_identities or no_identities
+
+    @pytest.mark.parametrize("hamiltonian", test_hamiltonians)
+    def test_observable_types(self, hamiltonian):
+        allowed_obs = (Tensor, Identity, PauliX, PauliY, PauliZ)
+
+        decomposed_coeff, decomposed_obs = pu.decompose_hamiltonian(hamiltonian)
+        assert all([isinstance(o, allowed_obs) for o in decomposed_obs])
```
[the remark this answers is not in the document]
Comment (nvitucci) [reply tests/test_utils.py:112]: Nice, didn't think of this!

**Comment by nvitucci** [comment]
Thanks @trbromley ! I slightly refactored the check on the shape along with the test, so now the exception is raised when the shape is not square too. I've also added the entry to the changelog.

File: doc/code/qml_utils.rst [hunk 9]
```diff
@@ -12,4 +12,4 @@ qml.utils
     :no-heading:
     :include-all-objects:
     :no-inheritance-diagram:
-    :skip: Iterable, Variable, OrderedDict
+    :skip: Iterable, Variable, OrderedDict, matmul
```
Comment (trbromley) [inline doc/code/qml_utils.rst:15]: Nice spot!

File: pennylane/utils.py [hunk 10]
```diff
@@ -19,16 +19,93 @@
...
+    (-1.0) [I0 I1]
+    + (-1.5) [X1]
+    + (-0.5) [Y1]
+    + (-1.0) [Z1]
+    + (-1.5) [X0]
+    + (-1.0) [X0 X1]
+    + (-0.5) [X0 Z1]
+    + (1.0) [Y0 Y1]
+    + (-0.5) [Z0 X1]
+    + (-0.5) [Z0 Y1]
```
Comment (trbromley) [inline pennylane/utils.py:70]:
```suggestion
    (-1.0) [I<Wires = [0]> I<Wires = [1]>]
    + (-1.5) [I<Wires = [0]> X<Wires = [1]>]
    + (-0.5) [I<Wires = [0]> Y<Wires = [1]>]
    + (-1.0) [I<Wires = [0]> Z<Wires = [1]>]
    + (-1.5) [X<Wires = [0]> I<Wires = [1]>]
    + (-1.0) [X<Wires = [0]> X<Wires = [1]>]
    + (-0.5) [X<Wires = [0]> Z<Wires = [1]>]
    + (1.0) [Y<Wires = [0]> Y<Wires = [1]>]
    + (-0.5) [Z<Wires = [0]> X<Wires = [1]>]
    + (-0.5) [Z<Wires = [0]> Y<Wires = [1]>]
```
I think this is due to the new `Wires` class

File: tests/test_utils.py [hunk 11]
```diff
@@ -63,6 +69,73 @@
...
+
+class TestDecomposition:
+    """Tests the decompose_hamiltonian function"""
+
+    @pytest.mark.parametrize("hamiltonian", [np.ones((3, 3)), np.ones((4, 2)), np.ones((2, 4))])
+    def test_wrong_shape(self, hamiltonian):
+        """Tests that an exception is raised if the Hamiltonian does not have
+        the correct shape"""
+        with pytest.raises(
+            ValueError, match=re.escape("The Hamiltonian should have shape"),
```
Comment (trbromley) [inline tests/test_utils.py:87]:
```suggestion
            ValueError, match="The Hamiltonian should have shape",
```

File: tests/test_utils.py [hunk 12]
```diff
@@ -15,15 +15,21 @@
 Unit tests for the :mod:`pennylane.utils` module.
 """
 # pylint: disable=no-self-use,too-many-arguments,protected-access
-import pytest
+import operator
+import functools
+import itertools
 from unittest.mock import MagicMock
+import pytest
+import re
```
Comment (trbromley) [inline tests/test_utils.py:23]:
```suggestion
```
(Also requires the suggestion below)

File: tests/test_utils.py [hunk 13]
```diff
@@ -15,15 +15,21 @@
 Unit tests for the :mod:`pennylane.utils` module.
 """
 # pylint: disable=no-self-use,too-many-arguments,protected-access
-import pytest
+import operator
```
Comment (trbromley) [inline tests/test_utils.py:18]:
```suggestion
```
I think this one isn't needed

File: pennylane/utils.py (hunk 10 above)
Comment (josh146) [reply pennylane/utils.py:70]: Note that this is a bug, and will be reverted in #684! So don't commit this suggestion, the original is how it should appear :)

Comment (trbromley) [reply pennylane/utils.py:70]: Good, the original is better :laughing:

**Pull request merged**